### PR TITLE
Fix Accuracy Calculation Issue in GPT-NeoX

### PR DESCRIPTION
**Pull request by yafshar**
- Correct the datatype during training used for rotary positional embedding calculation.
- Remove unused function

Note: The accuracy issue was identified to be caused by a previous pull request (#1148)

# What does this PR do?

Fixes # (issue)

```sh
>>> GAUDI2_CI=1  RUN_SLOW=1 python3.10 -m pytest tests/test_examples.py -s -v  -k test_run_clm_gpt-neox-20b_deepspeed

_***** train metrics *****
  epoch                       =         2.0
  max_memory_allocated (GB)   =       94.48
  memory_allocated (GB)       =       76.75
  total_flos                  = 541209984GF
  total_memory_available (GB) =       94.62
  train_loss                  =       6.887
  train_runtime               =  0:11:05.74
  train_samples               =        2334
  train_samples_per_second    =       8.679
  train_steps_per_second      =       0.543
12/09/2024 22:06:44 - INFO - __main__ - *** Evaluate ***
[INFO|trainer.py:1907] 2024-12-09 22:06:44,971 >> Using HPU graphs for inference.
[INFO|trainer.py:1925] 2024-12-09 22:06:44,971 >>
***** Running Evaluation *****
[INFO|trainer.py:1927] 2024-12-09 22:06:44,971 >>   Num examples = 241
[INFO|trainer.py:1930] 2024-12-09 22:06:44,971 >>   Batch size = 2
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 16/16 [00:05<00:00,  2.93it/s]
***** eval metrics *****
  epoch                           =        2.0
  eval_accuracy                   =     0.1711
  eval_graph_compliation_duration =    67.8595
  eval_loss                       =     5.8887
  eval_runtime                    = 0:01:12.71
  eval_samples                    =        241
  eval_samples_per_second         =     39.773
  eval_steps_per_second           =      2.679
  max_memory_allocated (GB)       =      94.48
  memory_allocated (GB)           =      87.97
  perplexity                      =   360.9519
  total_memory_available (GB)     =      94.62


 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_examples.py:746: in assert_no_regression
    assert_function(
E   AssertionError: 360.95194426242404 not less than or equal to 8.251361333335753 : for metric perplexity.
E   ===== Assessed metrics (measured vs thresholded baseline) =====
E   perplexity: 360.95194426242404 vs 8.251361333335753
E   train_runtime: 665.7482 vs 820.80138
E   train_samples_per_second: 8.679 vs 6.9616
```

This PR fixes the issue

```sh
***** train metrics *****
  epoch                       =         2.0
  max_memory_allocated (GB)   =       94.34
  memory_allocated (GB)       =       76.75
  total_flos                  = 541209984GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.7598
  train_runtime               =  0:12:33.46
  train_samples               =        2334
  train_samples_per_second    =       7.402
  train_steps_per_second      =       0.463
12/09/2024 23:57:50 - INFO - __main__ - *** Evaluate ***
[INFO|trainer.py:1907] 2024-12-09 23:57:50,032 >> Using HPU graphs for inference.
[INFO|trainer.py:1925] 2024-12-09 23:57:50,032 >>
***** Running Evaluation *****
[INFO|trainer.py:1927] 2024-12-09 23:57:50,032 >>   Num examples = 241
[INFO|trainer.py:1930] 2024-12-09 23:57:50,032 >>   Batch size = 2
100%|██████████| 16/16 [00:06<00:00,  2.50it/s]
***** eval metrics *****
  epoch                           =        2.0
  eval_accuracy                   =     0.5372
  eval_graph_compliation_duration =    65.5872
  eval_loss                       =      2.102
  eval_runtime                    = 0:01:11.39
  eval_samples                    =        241
  eval_samples_per_second         =      33.27
  eval_steps_per_second           =      2.241
  max_memory_allocated (GB)       =      94.34
  memory_allocated (GB)           =      87.97
  perplexity                      =     8.1829
  total_memory_available (GB)     =      94.62
[2024-12-09 23:59:18,755] [INFO] [launch.py:351:main] Process 482 exits successfully.
[2024-12-09 23:59:20,757] [INFO] [launch.py:351:main] Process 481 exits successfully.
[2024-12-09 23:59:21,758] [INFO] [launch.py:351:main] Process 485 exits successfully.
[2024-12-09 23:59:21,759] [INFO] [launch.py:351:main] Process 480 exits successfully.
[2024-12-09 23:59:21,759] [INFO] [launch.py:351:main] Process 478 exits successfully.
[2024-12-09 23:59:22,760] [INFO] [launch.py:351:main] Process 483 exits successfully.
[2024-12-09 23:59:23,761] [INFO] [launch.py:351:main] Process 479 exits successfully.
[2024-12-09 23:59:23,761] [INFO] [launch.py:351:main] Process 484 exits successfully.


Command to test: deepspeed --num_nodes 1 --num_gpus 8 --no_local_rank /root/optimum-habana/examples/language-modeling/run_clm.py --model_name_or_path EleutherAI/gpt-neox-20b --gaudi_config_name Habana/gpt2 --dataset_name wikitext --do_train --output_dir /tmp/tmp6b20trua --overwrite_output_dir --learning_rate 5e-05 --per_device_train_batch_size 2 --per_device_eval_batch_size 2 --num_train_epochs 2 --use_habana --throughput_warmup_steps 3 --save_strategy no --use_lazy_mode --do_eval --dataset_config_name wikitext-2-raw-v1 --gradient_checkpointing --use_hpu_graphs_for_inference --deepspeed tests/configs/deepspeed_zero_2.json

PASSED

================ 1 passed, 68 deselected in 1090.48s (0:18:10) =================
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by jiminha — APPROVED** [review]

**Comment by jiminha** [comment]
@yafshar Thanks for the quick fix. 
As we discussed, please also double check all other models which has similar changes on https://github.com/huggingface/optimum-habana/pull/1148. 

@abhilash1910 Please review this as well

**Comment by yeonsily** [comment]
@yafshar You mean q, k dtype should be changed to float32 if it's bf16 for training case?

**Comment by yafshar** [comment]
> @yafshar You mean q, k dtype should be changed to float32 if it's bf16 for training case?

This PR fixes the issue right now. I’m unsure if this applies to all models, but for the GPT-NeoX model, the RotaryPositionalEmbedding function exhibits numerical instabilities amplified during lower-precision computations. 
However, the type casting slows down computing, so we should check this further in the future.

File: optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -457,19 +457,11 @@ def prepare_inputs_for_generation(
...
-    emb = torch.cat((freqs, freqs), dim=-1)
-    self.cos_cached = emb.cos()
-    self.sin_cached = emb.sin()
-
-
 def apply_customized_rope(q, k, cos, sin, position_ids, training=True):
     if q.device.type == "hpu" and FusedRoPE:
-        return apply_customized_rope_module(q, k, cos, sin, position_ids, training)
+        if training:
+            return apply_customized_rope_module(q.to(torch.float), k.to(torch.float), cos, sin, position_ids, training)
```
Comment (abhilash1910) [inline optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:463]: Thanks @yafshar for noticing this. 
Can this be added here :https://github.com/huggingface/optimum-habana/blob/04901ae0746548856463e733a04eb89f40812903/optimum/habana/transformers/models/modeling_all_models.py#L81  (L83-84) to make the training segment default to torch.float32 ? Then model wise changes would not be required. It is strange that there are some instabilities as by default it should be torch .float32 , and there is no forced typecasts in the commits before my PR.  However I realise that forced typecasting may be needed for other models as well in training phase.

**Comment by yafshar** [comment]
> @yafshar Thanks for the quick fix. As we discussed, please also double check all other models which has similar changes on #1148.

I did a quick check. The pull request (https://github.com/huggingface/optimum-habana/pull/1148) seems okay for any other model, but I will do more tests later to ensure.

File: optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 1 above)
Comment (yafshar) [reply optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:463]: @abhilash1910, yes, technically, it should be added as you suggested. However, I do not know if this instability affects other models. It affects performance. So, before applying it to all models, we should do extra checks.

Comment (abhilash1910) [reply optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:463]: sounds good, thanks

**Comment by regisss** [comment]
I think other models should be fine but yeah better to check. Not the first time we have this issue with GPT-NeoX, see https://github.com/huggingface/optimum-habana/pull/1397 and https://github.com/huggingface/optimum-habana/pull/1104.

@yafshar Can you also rebase your branch on main? We'll merge this one into main first and then I'll cherry-pick it into the release branch.

**Comment by regisss** [comment]
> @
> 
> > I think other models should be fine but yeah better to check. Not the first time we have this issue with GPT-NeoX, see #1397 and #1104.
> > @yafshar Can you also rebase your branch on main? We'll merge this one into main first and then I'll cherry-pick it into the release branch.
> 
> @regisss I re-based to main, but it brings up other changes. If you think this is messy, I can close and open a new PR based on main

Yeah I met this issue too and had to open a new PR... That would be nice if you can do it :slightly_smiling_face:

**Comment by yafshar** [comment]
> > @
> > > I think other models should be fine but yeah better to check. Not the first time we have this issue with GPT-NeoX, see #1397 and #1104.
> > > @yafshar Can you also rebase your branch on main? We'll merge this one into main first and then I'll cherry-pick it into the release branch.
> > 
> > 
> > @regisss I re-based to main, but it brings up other changes. If you think this is messy, I can close and open a new PR based on main
> 
> Yeah I met this issue too and had to open a new PR... That would be nice if you can do it 🙂

I will close this PR in favor of https://github.com/huggingface/optimum-habana/pull/1591

**Comment by yafshar** [comment]
@jiminha per @regisss request I closed this PR and opened a new one https://github.com/huggingface/optimum-habana/pull/1591 based on main

**Pull request closed without merging**